### PR TITLE
Wu Anti-aliased lines demo, first draft.

### PR DESCRIPTION
**Pull request by calroc**
I don't think this is ready to go into the `graphics.h` header file but I wanted to show what I've got so far.

I wrote four "reflections" of the basic Wu algorithm and a wrapper to call the right one.  (For me it's easier to do this than to figure out a single version that correctly handles the four cases, and then writing the single version is easier from the boilerplate.)

I embedded the line-drawing code in a simple demo.

Special-case code for horizontal, vertical and 45° degree lines is omitted for now.

### CODE REVIEW

File: ncc/examples/wu.c [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        if (error_accumulator > LIMIT)
+        {
+            error_accumulator = error_accumulator - LIMIT;
+            ++y1;
+            --y0;
+        }
+    }
+}
+
+void draw_Wu_line_fourth_octant(u32* fb, u32 fb_width, u32 fb_height, u32 x0, u32 y0, u32 x1, u32 y1, u32 color)
```
Comment (maximecb) [inline ncc/examples/wu.c:190]: nitpick but `_wu_` should be lowercase to match snake-case, even though it's a proper name.

File: ncc/examples/wu.c [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+size_t FRAME_HEIGHT = 400;
+
+u32 frame_buffer[160000];
+
+// Current mouse pointer position
+size_t pos_x = 200;
+size_t pos_y = 200;
+
+void
+carefree_alpha_blend_plot_pixel(u32* dest, size_t dest_stride, u64 x, u64 y, u32 color, u8 alpha)
```
Comment (maximecb) [inline ncc/examples/wu.c:18]: Here I don't understand why `carefree` :sweat_smile:

Do some versions of alpha blending worry more than others :thinking:

**Comment by maximecb** [comment]
I was able to test out the demo locally. Looks good! :)

I'm looking forward to being able to draw a 3D spinning cube soon hehehe.

**Comment by maximecb** [comment]
I used your test code and realized that the implementation of Bresenham's algorithm I had included in `include/graphics.h` was completely broken, and found a bug in the compiler too :sweat_smile: 

I just pushed a fix for comparison operators in ncc. It wasn't doing comparison of negative `int` values correctly. This may have affected your code. In general, need to be careful if comparing signed and unsigned values, e.g. `int` and `u32`. If you have a signed and an unsigned value, then an unsigned comparison operator will be used, which doesn't account for negative numbers.

If you want a signed comparison, you need to either use two `int` values, or cast to int first, e.g.

```
int a;
u32 b;
if (x < b) {} // uses an unsigned comparison
if (a < (int)b) {} // uses a signed comparison that accounts for negative values correctly
```

File: ncc/examples/wu.c (hunk 2 above)
Comment (calroc) [reply ncc/examples/wu.c:18]: It's just to remind myself that this function doesn't check it's arguments, it will just blithely write to any old address if you give it bad args.  The image in my mind is the scene from *Austin Powers* where Dr. Evil is like, "We'll just close the doors and assume everything went to plan." :)

**Comment by calroc** [comment]
> I was able to test out the demo locally. Looks good! :)

Ah, that's so cool!  I wanted to have something simple that obviously showed that lines in all four quadrants worked and looked okay, and that was at least a little interactive and visually interesting.  I'm not a graphics designer or demo scene writer, eh?

> I'm looking forward to being able to draw a 3D spinning cube soon hehehe.

I remembered that I had uploaded my 3D demo to github years ago: https://github.com/calroc/Tkinter3D  (I don't recall whether I ever updated it to Py3.)  It's a really direct translation of the (C++?) code in "Physics For Game Developers" ( https://archive.org/details/physics-for-game-developers-2e )  https://github.com/calroc/Tkinter3D/blob/master/math3d.py

I understood it when I wrote it but now I'm a little fuzzy.  In any event, it's all there: 3D math, a simple scene graph, and the frustum transform for perspective.  Like I said, I got to the point of animating a rotating cube (dots on a Tk Canvas).  I always meant to use this (or at least what I learned from it) to implement a UI.  That's how long I've been waiting for UVM! :)  The code was old when I put it in github.

In any event you're welcome to it.  I'm probably going to spend at least a few days messing around with cheap fake 3D effects using integers and like tables of pre-computed *sin* values or something.  I don't actually need full 3D for what I want to do (affine transforms should cover most of it) but I too would like to see cool wireframe flythough demos and such.  :)  So if you don't get to it before me I'll eventually do it. ;)

**Comment by calroc** [comment]
I clicked the wrong button (I promise I won't mention it more than once a week, but I do not like GitHub's UI.)

File: ncc/examples/wu.c (hunk 2 above)
Comment (maximecb) [reply ncc/examples/wu.c:18]: Should probably use the term `unsafe_`

**Comment by maximecb** [comment]
> You're welcome to it. I'm probably going to spend at least a few days messing around with cheap fake 3D effects using integers and like tables of pre-computed sin values or something. I don't actually need full 3D for what I want to do (affine transforms should cover most of it) but I too would like to see cool wireframe flythough demos and such. :) So if you don't get to it before me I'll eventually do it. ;)

I'm curious to see it if you implement it. Having fun 3D demos to show people would definitely make UVM more attractive.

For this PR, are you going to combine the 4 functions into one as you had hinted?

**Comment by calroc** [comment]
> For this PR, are you going to combine the 4 functions into one as you had hinted?

I could, but is that desirable?  To my way of thinking it seems better to have these four fundamental "reflections" with a router or dispatcher function, because if you happen to know up front which one to use for your lines you can avoid the overhead of dispatching and just call them directly.  E.g. for a rotating cube most of the lines most of the frames will be in the same octant, one could keep the lines in separate per-octant lists (and either move them from one to the other as needed or something?)  The basic idea is to keep sorting separate from the pixel-pushing.

> I'm curious to see it if you implement it. Having fun 3D demos to show people would definitely make UVM more attractive.

I'll get to it eventually if no one else does first, I'll have to at least skim that book again.

File: ncc/examples/wu.c [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        }
+        else
+        {
+            //draw_vertical_line
+        }
+    }
+    else if (y1 == y0)  // and we know x1 != x0
+    {
+        //draw_horizontal_line
+    }
```
Comment (maximecb) [inline ncc/examples/wu.c:58]: Is this case missing from the implementation?

Comment (calroc) [reply ncc/examples/wu.c]: Oh right!  I'll add the special cases tonight.

Comment (calroc) [reply ncc/examples/wu.c]: Hey, I think I should be able to get to this today.  (It's been an eventful week.)

Comment (maximecb) [reply ncc/examples/wu.c]: Nice. Please make sure to test all the edge cases :)

Comment (calroc) [reply ncc/examples/wu.c]: I haven't forgotten about this PR, I've just been really busy.  I'll get back to it soon.

The last time I sync'd my branch to main and ran the demo there was some problem with the colors.  I haven't yet had time to revisit that and figure out what's wrong.  Hopefully it just goes away after then next pull, eh?

That reminds me, the font demo had a few spurious pixels above the first glyph.  I never figured out what that was about, and now it's moot.  Such is life, eh?

Comment (maximecb) [reply ncc/examples/wu.c]: > I haven't forgotten about this PR, I've just been really busy. I'll get back to it soon.

No stress :)

If you can find a small repro wrt color issues, open an issue and I'll take a look 👍

Comment (calroc) [reply ncc/examples/wu.c]: Hey, I've been busy, but I took a moment to look at this today and I found out what was causing the messed up colors.  This change fixed it (`red`, `green`, and `blue` are `u8`):

``` C
-    *pix_ptr = (red << 16) | (green << 8) | blue;
+    *pix_ptr = rgb32(red, green, blue);
```

**Comment by maximecb** [comment]
Hi @calroc ! Is this ready with all the edge cases handled and tested? :)

**Comment by calroc** [comment]
Hi!  Yeah, it all seems to work: horizontal, vertical, and 45° lines.  When the program starts the initial (x,y) location for the mouse results in a nice overlap between the two sets of diagonals and which shows off the anti-aliasing nicely.

It needs some editing for style and cleanup (`s/carefree/unsafe/`, etc.) still.  I should be able to get to that today.  (It's been raining all month here and I'm way behind on setting up my garden for Spring!)

**Comment by maximecb** [comment]
Alright, almost done then :)
